pageserver/src/tenant.rs

```diff
@@ -2812,7 +2812,50 @@ impl Tenant {
         cancel: &CancellationToken,
         ctx: &RequestContext,
     ) -> anyhow::Result<Vec<Arc<Timeline>>> {
-        // grab mutex to prevent new timelines from being created here.
+        let timelines = self
+            .timelines
+            .lock()
+            .unwrap()
+            .values()
+            .filter(|tl| match target_timeline_id.as_ref() {
+                Some(required) => &tl.timeline_id == required,
+                None => true,
+            })
+            .cloned()
+            .collect::<Vec<_>>();
+
+        let mut gc_cutoffs: HashMap<TimelineId, GcCutoffs> =
+            HashMap::with_capacity(timelines.len());
+
+        // before taking the gc_cs lock, do the heavier weight finding of gc_cutoff points for
+        // currently visible timelines.
+        for timeline in timelines.iter() {
+            let cutoff = timeline
+                .get_last_record_lsn()
+                .checked_sub(horizon)
+                .unwrap_or(Lsn(0));
+
+            // FIXME: using questionmark here is wrong: either the whole tenant is shutting down,
+            // or a single timeline is getting deleted, we should at most log the error
+            let res = timeline.find_gc_cutoffs(cutoff, pitr, cancel, ctx).await;
```

Comment:
I see no questionmark.

Edit: ah, there was one in the removed code. Generally, referring to removed code in a PR in this way is just confusing to readers, please find an alternative formulation.

Reply:
Ah yes, forgot to remove the FIXME with the addition of synthetic size tests.

Reply:
[e8a0354](https://github.com/neondatabase/neon/pull/7585/commits/e8a0354787a63a37312840028bfa81530f772864)